L1Trigger/TrackFindingTracklet/src/KFin.cc

```diff
@@ -0,0 +1,424 @@
+      const TTBV sectorPhi(
+          dataFormats_->format(Variable::sectorPhi, Process::kfin).ttBV(track->sector_ / setup_->numSectorsEta()));
+      const TTBV sectorEta(
+          dataFormats_->format(Variable::sectorEta, Process::kfin).ttBV(track->sector_ % setup_->numSectorsEta()));
+      const TTBV inv2R(dataFormats_->format(Variable::inv2R, Process::kfin).ttBV(track->inv2R_));
+      const TTBV phiT(dataFormats_->format(Variable::phiT, Process::kfin).ttBV(track->phiT_));
+      const TTBV cot(dataFormats_->format(Variable::cot, Process::kfin).ttBV(track->cot_));
+      const TTBV zT(dataFormats_->format(Variable::zT, Process::kfin).ttBV(track->zT_));
+      return FrameTrack(track->ttTrackRef_,
+                        Frame("1" + maybe.str() + sectorPhi.str() + sectorEta.str() + phiT.str() + inv2R.str() +
+                              zT.str() + cot.str()));
+    };
+    auto frameStub = [this](Track* track, int layer) {
+      auto equal = [layer](Stub* stub) { return stub->valid_ && stub->layer_ == layer; };
+      const auto it = find_if(track->stubs_.begin(), track->stubs_.end(), equal);
+      if (it == track->stubs_.end() || !(*it)->valid_)
+        return FrameStub();
+      Stub* stub = *it;
+      const TTBV r(dataFormats_->format(Variable::r, Process::kfin).ttBV(stub->r_));
+      const TTBV phi(dataFormats_->format(Variable::phi, Process::kfin).ttBV(stub->phi_));
+      const TTBV z(dataFormats_->format(Variable::z, Process::kfin).ttBV(stub->z_));
+      const TTBV dPhi(dataFormats_->format(Variable::dPhi, Process::kfin).ttBV(stub->dPhi_));
+      const TTBV dZ(dataFormats_->format(Variable::dZ, Process::kfin).ttBV(stub->dZ_));
+      return FrameStub(stub->ttStubRef_, Frame("1" + r.str() + phi.str() + z.str() + dPhi.str() + dZ.str()));
+    };
+    auto invalid = [](Track* track) { return track && !track->valid_; };
+    auto acc = [invalid](int& sum, Track* track) { return sum += (invalid(track) ? 1 : 0); };
+    const int offsetTrack = region_ * channelAssignment_->numChannelsTrack();
+    for (int channel = 0; channel < channelAssignment_->numChannelsTrack(); channel++) {
+      const int channelTrack = offsetTrack + channel;
+      const int offsetStub = channelTrack * setup_->numLayers();
+      vector<Track*>& input = input_[channel];
+      // fill lost tracks and stubs without gaps
+      const int lost = accumulate(input.begin(), input.end(), 0, acc);
+      lostTracks[channelTrack].reserve(lost);
+      for (int layer = 0; layer < setup_->numLayers(); layer++)
+        lostStubs[offsetStub + layer].reserve(lost);
+      for (Track* track : input) {
+        if (!track || track->valid_)
+          continue;
+        lostTracks[channelTrack].emplace_back(frameTrack(track));
+        for (int layer = 0; layer < setup_->numLayers(); layer++)
+          lostStubs[offsetStub + layer].emplace_back(frameStub(track, layer));
+      }
+      // fill accepted tracks and stubs with gaps
```

Comment:
what are stubs with "gaps"?

Reply:
I meant (track and stubs) with gaps. a gap is a clock tick where the algorithm can't provide valid output data.